### PR TITLE
SDKS-2413 Android Key Attestation Implementation

### PR DESCRIPTION
**Pull request by witrisna**
# JIRA Ticket

[SDKS-2413](https://bugster.forgerock.org/jira/browse/SDKS-2413)

# Description

Android Key Attestation Implementation

# Definition of Done Checklist:

- [ ] Acceptance criteria is met.
- [ ] All tasks listed in the user story have been completed.
- [ ] Coded to standards.
- [ ] Ensure backward compatibility.
- [ ] API reference docs is updated.
- [ ] Unit tests are written.
- [ ] Integration tests are written.
- [ ] e2e tests are written.
- [ ] Functional spec is written/updated.
- [ ] contains example code snippets.
- [ ] Change log updated.
- [ ] Documentation story is created and tracked.
- [ ] Tech debts and remaining tasks are tracked in separated ticket(s).

### CODE REVIEW

File: forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/ApplicationPinDeviceAuthenticator.kt [hunk 1]
```diff
@@ -49,7 +50,10 @@ open class ApplicationPinDeviceAuthenticator(private val pinCollector: PinCollec
 
     private val worker = Executors.newSingleThreadScheduledExecutor()
 
-    override suspend fun generateKeys(context: Context): KeyPair {
+    override suspend fun generateKeys(context: Context, attestation: Attestation): KeyPair {
```
Comment (jeyanthanperiyasamy) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/ApplicationPinDeviceAuthenticator.kt:53]: we dont support attestation for Application PIN ? may be we need to mention in the docs

File: forgerock-auth/src/main/java/org/forgerock/android/auth/callback/Attestation.kt [hunk 2]
```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright (c) 2023 ForgeRock. All rights reserved.
+ *
+ *  This software may be modified and distributed under the terms
+ *  of the MIT license. See the LICENSE file for details.
+ */
+
+package org.forgerock.android.auth.callback
+
+sealed class Attestation(val challenge: ByteArray? = null) : java.io.Serializable {
```
Comment (jeyanthanperiyasamy) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/callback/Attestation.kt:10]: i believe this is for the byteArray this has to Serializable ?

File: forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceSigningVerifierCallback.kt
Comment (jeyanthanperiyasamy) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceSigningVerifierCallback.kt:1]: So the DeviceSigningVerifierCallback does not require to do any attestation . the only time we attest is when we generate the keys in DeviceBindingCallBack ?

**Review by jeyanthanperiyasamy — COMMENTED** [review]
added few questions for clarification. overall make sense

File: forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/ApplicationPinDeviceAuthenticator.kt (hunk 1 above)
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/ApplicationPinDeviceAuthenticator.kt:53]: Yes, the following limitation will be required to document

- Key Attestation only support for Android version > 24 (Android N), for device running Android version 23, Key Attestation verification will be skipped.
- Key Attestation is not supported for Application Pin Authentication Type
- Key Attestation is not supported for iOS device

Those are documented in the Design specification and will need to update in the official documentation.

File: forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceSigningVerifierCallback.kt
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceSigningVerifierCallback.kt:1]: Correct, we only attest the key when we create the key, but not using the key.

**Review by jeyanthanperiyasamy — APPROVED** [review]

**Review by spetrov — APPROVED** [review]
LGTM 👍🏻

**Pull request merged**